src/cmd/sanity_test/restore/exchange.go

```diff
@@ -120,92 +55,71 @@ func verifyEmailData(ctx context.Context, restoreMessageCount, messageCount map[
 	}
 }
 
-// getAllSubFolder will recursively check for all subfolders and get the corresponding
-// email count.
-func getAllMailSubFolders(
+func buildSanitree(
 	ctx context.Context,
 	ac api.Client,
-	testUser string,
-	r models.MailFolderable,
-	parentFolder,
-	dataFolder string,
-	messageCount map[string]int32,
-) {
-	folderID := ptr.Val(r.GetId())
-	ctx = clues.Add(ctx, "parent_folder_id", folderID)
-
-	childFolders, err := ac.Mail().GetContainerChildren(ctx, testUser, folderID)
+	userID, folderName string,
+) *common.Sanitree[models.MailFolderable] {
+	gcc, err := ac.Mail().GetContainerByName(
+		ctx,
+		userID,
+		api.MsgFolderRoot,
+		folderName)
 	if err != nil {
-		common.Fatal(ctx, "getting mail subfolders", err)
+		common.Fatal(
+			ctx,
+			fmt.Sprintf("finding folder by name %q", folderName),
+			err)
 	}
 
-	for _, child := range childFolders {
-		var (
-			childDisplayName = ptr.Val(child.GetDisplayName())
-			childFolderCount = ptr.Val(child.GetChildFolderCount())
-			//nolint:forbidigo
-			fullFolderName = stdpath.Join(parentFolder, childDisplayName)
-		)
-
-		if filters.PathContains([]string{dataFolder}).Compare(fullFolderName) {
-			messageCount[fullFolderName] = ptr.Val(child.GetTotalItemCount())
-			// recursively check for subfolders
-			if childFolderCount > 0 {
-				parentFolder := fullFolderName
-
-				getAllMailSubFolders(
-					ctx,
-					ac,
-					testUser,
-					child,
-					parentFolder,
-					dataFolder,
-					messageCount)
-			}
-		}
+	mmf, ok := gcc.(models.MailFolderable)
+	if !ok {
+		common.Fatal(
+			ctx,
+			"mail folderable required",
+			clues.New("casting "+*gcc.GetDisplayName()+" to models.MailFolderable"))
+	}
+
+	root := &common.Sanitree[models.MailFolderable]{
```

Comment:
I'm guessing the idea is to expand it to other services later.

Reply:
Yup!  Spot on.